File: metaprogramming/metaprogramming.java
import java.util.*;

public class metaprogramming{
   public static void main(String[] args) {
      HashMap<String, Integer> values  = new HashMap<String, Integer>();
      Scanner in = new Scanner(System.in);

      String line;
      while(in.hasNextLine()){
         line = in.nextLine();
         String[] split = line.split(" ");
         if(split[0].equals("define")){
            values.put(split[2],Integer.parseInt(split[1]));
            
         }
         else if(split[0].equals("eval")){
            if (!values.containsKey(split[1]) || !values.containsKey(split[3])){
               System.out.println("undefined");
            }
            else{
               if (split[2].equals("=")){
                  System.out.println(values.get(split[1])  == values.get(split[3]));
               }
               else if(split[2].equals("<")){
                  System.out.println(values.get(split[1]) < values.get(split[3]));
               }
               else if(split[2].equals(">")){
                  System.out.println(values.get(split[1]) > values.get(split[3]));
               }
               else{
                  System.out.println("error");
               }
            }
         }
      }  
   }
}

File: metaprogramming/metaprogramming.py
import sys

var = dict()

for line in sys.stdin.readlines():
    p = line.split()
    if p[0] == 'define':
        var[p[2]] = int(p[1])
    else:
        if p[1] not in var or p[3] not in var:
            print('undefined')
            continue
        first = var[p[1]]
        second = var[p[3]]
        op = p[2]
        if op == '=':
            print('true' if first == second else 'false')
        elif op == '>':
            print('true' if first > second else 'false')
        elif op == '<':
            print('true' if first < second else 'false')
